cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/k8s/UserServiceProgramMain.java

```diff
@@ -0,0 +1,379 @@
+            }
+          });
+
+          bind(ExploreClient.class).to(UnsupportedExploreClient.class);
+          bind(OwnerAdmin.class).to(NoOpOwnerAdmin.class);
+        }
+      }
+    );
+  }
+
+  @Override
+  protected void addServices(Injector injector, List<? super com.google.common.util.concurrent.Service> services,
+                             List<? super AutoCloseable> closeableResources) {
+    services.add(injector.getInstance(MetricsCollectionService.class));
+  }
+
+  @Nullable
+  @Override
+  protected LoggingContext getLoggingContext() {
+    return null;
+  }
+
+  @Override
+  protected ServiceOptions createOptions() {
+    return new ServiceOptions();
+  }
+
+  @Override
+  public void init(String[] args) throws MalformedURLException {
+    super.init(args);
+
+    ApplicationSpecification appSpec = readFile(new File(options.getAppSpecPath()), ApplicationSpecification.class);
+    programOptions = readFile(new File(options.getProgramOptionsPath()), ProgramOptions.class);
+    programRunId = programOptions.getProgramId().run(ProgramRunners.getRunId(programOptions));
+
+    controllerFuture = new CompletableFuture<>();
+    programCompletion = new CompletableFuture<>();
+    maxStopSeconds = cConf.getLong(Constants.AppFabric.PROGRAM_MAX_STOP_SECONDS, 60);
+
+    programRunner = injector.getInstance(ServiceProgramRunner.class);
+
+    // fetch the app artifact
+    ArtifactFinder artifactFinder = injector.getInstance(ArtifactFinder.class);
+
+    NamespaceId programNamespace = programRunId.getNamespaceId();
+    Set<co.cask.cdap.api.artifact.ArtifactId> pluginArtifacts = new HashSet<>();
+    for (Plugin plugin : appSpec.getPlugins().values()) {
+      pluginArtifacts.add(plugin.getArtifactId());
+    }
+
+    // need to set the plugins directory in system args, as well as the host to bind to
+    File pluginsDir = new File(options.getTmpDir(), "plugins");
+    Map<String, String> systemArgs = new HashMap<>(programOptions.getArguments().asMap());
+    systemArgs.put(ProgramOptionConstants.PLUGIN_DIR, pluginsDir.getAbsolutePath());
+    systemArgs.put(ProgramOptionConstants.HOST, options.getBindAddress());
+    programOptions = new SimpleProgramOptions(programRunId.getParent(), new BasicArguments(systemArgs),
+                                              programOptions.getUserArguments());
+
+    try {
+      File programDir = new File(options.getTmpDir(), "program");
+      DirUtils.mkdirs(programDir);
+      File programJarFile = new File(programDir, "program.jar");
+      Location programJarLocation =
+        artifactFinder.getArtifactLocation(convert(programNamespace, appSpec.getArtifactId()));
+      Files.copy(programJarLocation.getInputStream(), programJarFile.toPath());
+
+      DirUtils.mkdirs(pluginsDir);
+      for (co.cask.cdap.api.artifact.ArtifactId artifactId : pluginArtifacts) {
+        Location location = artifactFinder.getArtifactLocation(convert(programNamespace, artifactId));
```

Comment:
I think it's better for the `getArtifactLocation` method be a batch method such that it can lookup multiple artifact locations in one call. It also help parallelizing the file copy in future if needed.